### DISCUSSION TITLE
Specify Ports for resources created by DistributedApplicationTestingBuilder

### DISCUSSION THREAD

**Discussion by gdodd1977** [microsoft/aspire#4501] [Q&A] [answered]
I've created an Aspire-Xunit integration test project that is doing the following to create an app host off my Aspire project:
` var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.CopilotFlow_LocalHost_StartHere>();
 await using var app = await appHost.BuildAsync();`

The Aspire project starts up three services:
`builder.AddProject<Projects.CopilotFlow_LocalHelper_Service>($"{targetEnvironment}LocalHelperService");
builder.AddProject<Projects.Microsoft_PowerAutomate_CopilotFlow_Web_Service>($"{targetEnvironment}WebService");
builder.AddProject<Projects.Microsoft_PowerAutomate_CopilotFlow_Worker_Service>($"{targetEnvironment}WorkerService");`

The issue is that the WebService requires a call to the LocalHelper_Service at a specific configured port. i.e. localhost:1234. But the code in the integration test using the DistributedApplicationTestingBuilder assigns a random port to the LocalHelper_Service.

So I have two questions.
1. Why wouldn't the launchSettings of the LocalHelper_Service be used when it's getting started, thus giving it the same port every time.
2. If that's not possible, how can I make sure that resource gets the right port every time DistributedApplicationTestingBuilder creates the app host?

**Comment by davidfowl** [comment]
> Why wouldn't the launchSettings of the LocalHelper_Service be used when it's getting started, thus giving it the same port every time.

To make sure you can run tests concurrently and to avoid port usage issues when tests start and stop.

> If that's not possible, how can I make sure that resource gets the right port every time DistributedApplicationTestingBuilder creates the app host?

If you don't hard code ports but instead resolve it from the resource it should work. Are you hard coding the port somewhere in your application?

You can turn off port randomization as well, but you probably also want to turn off test concurrency.

cc @ReubenBond

**Reply by gdodd1977** [reply]
No hard-coding per se, but we have appsettings.Local that has the localhost port we expect. So when running locally, CopilotFlow_Web_Service expects LocalHelper_Service to have a given port. So the randomizing of the port of LocalHelper_Service causes the call to fail when the test is running. 

Do you have some links on how to do either:

"If you don't hard code ports but instead resolve it from the resource it should work."

or 

"You can turn off port randomization as well, but you probably also want to turn off test concurrency."

**Reply by davidfowl** [reply]
> "If you don't hard code ports but instead resolve it from the resource it should work."

If you're using IConfiguration, you can set an environment variable in the apphost that overrides the port in the file. Can you share an example of what that configuration looks like?

**Reply by gdodd1977** [reply]
Which configuration? The one in my main aspire app or the one in my aspire xunit project? 

This is how I'm setting up the test project configuration:

`var config = new ConfigurationBuilder()
    .AddEnvironmentVariables()
    .AddJsonFile("appsettings.json")
    .Build();

hostBuilder
    .ConfigureHostConfiguration(builder => builder.AddConfiguration(config))
    .ConfigureServices(serviceCollection =>
    {
        var containerStartup = new ContainerStartup(config);
        containerStartup.ConfigureServices(serviceCollection);
    });`

**Reply by davidfowl** [reply]
The apphost project and the configuration.

**Reply by gdodd1977** [reply]
Not entirely sure what you mean by configuration. But our apphost project just has launchSettings and program.cs. This is program.cs:

![image](https://github.com/dotnet/aspire/assets/29410398/7b4e43c1-4c5f-412c-a14b-5914573828c9)

**Reply by davidfowl** [reply]
> Not entirely sure what you mean by configuratio
> No hard-coding per se, but we have appsettings.Local that has the localhost port we expect

This _appsettings.Local that has the localhost port we expect_

**Reply by gdodd1977** [reply]
![image](https://github.com/dotnet/aspire/assets/29410398/105a0ad6-6ddf-4483-860a-aba928e342e9)

This is in the WebService which calls into the LocalHelper_Service during local runs. This is the port it expects LocalHelper_Service to be running under locally.

**Reply by davidfowl** [reply]
You can set environment variables on your projects from the apphost so that those settings are overridden at runtime.

**Reply by gdodd1977** [reply]
Ok. Is there an example of that somewhere and how I would setup my xunit project to work with that?

**Reply by davidfowl** [reply]
From your apphost project, which service maps to ApiHubsClientConfiguration and FlowRpClientConfiguration?

**Reply by gdodd1977** [reply]
builder.AddProject<Projects.Microsoft_PowerAutomate_CopilotFlow_Web_Service>($"{targetEnvironment}WebService");

That's the one that has that config above and makes the call to those two clients.

**Reply by davidfowl** [reply, marked as answer]
```C#
var endpoint = project.GetEndpoint("http");

var localService = builder.AddProject<Projects.CopilotFlow_LocalHelper_Service>($"{targetEnvironment}LocalHelperService"); 

var httpEndpoint =  localService.GetEndpoint("http");

builder.AddProject<Projects.Microsoft_PowerAutomate_CopilotFlow_Worker_Service>($"{targetEnvironment}WorkerService");

builder.AddProject<Projects.Microsoft_PowerAutomate_CopilotFlow_Web_Service>($"{targetEnvironment}WebService")
           .WithEnvironment("ApiHubsClientConfiguration__ServiceUrl", httpEndpoint)
           .WithEnvironment("ApiHubsClientConfiguration__ServiceUrl", httpEndpoint);
```

**Reply by gdodd1977** [reply]
Thanks. I'll give that a shot.

**Reply by gdodd1977** [reply]
Ok. So I have another question. If I'm creating that new Env variable in the Web_Service, do I need to then use that to set my URL instead of the setting in my appsettings.local?

![image](https://github.com/dotnet/aspire/assets/29410398/3d377f7f-9273-47ed-b15e-5acd613bbfaf)

**Reply by gdodd1977** [reply]
I got this figured out. Thanks for the help!

**Reply by davidfowl** [reply]
Just to make sure I close the loop on this discussion. You don't have to change your application, the IConfiguration system translated environment variables to configuration entries. That format I have above e.g. `ApiHubsClientConfiguration__ServiceUrl` maps to `ApiHubsClientConfiguration:ServiceUrl`, which is the same as your JSON configuration.

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0#naming-of-environment-variables

**Reply by ejsmith** [reply]
@davidfowl  I have a similar problem, but I'm assigning fixed ports to my SQL Server container and it works as expected when just running the app, but when I run tests, it changes the port for SQL Server to a random port which makes it hard to inspect the DB when running tests since you don't know what the random port is.

**Reply by davidfowl** [reply]
Are you hard coding the port somewhere in your code or config? If you pass it in, you'll always get the right port.

**Reply by ejsmith** [reply]
Yes, I'm using `var sqlserver = builder.AddSqlServer("Sql", sqlPassword, port: 1433)` because I want the DB to run on a known port so I can connect to it while the app is running or I'm debugging a test and look at data. It respects that setting when running the app, but randomizes it when using `DistributedApplicationTestingBuilder`

**Reply by davidfowl** [reply]
You can hardcode the port there but the application shouldn't hard code it. Just pass it into the application as a reference and it'll flow correctly.

**Reply by ejsmith** [reply]
The application doesn't hard code it. My tests run fine even with the port being changed. That's not the point, I want the port to be known during testing so that I can hit a breakpoint and go look at the data. I don't understand why it's fixed for running the app, but random for testing.

**Reply by davidfowl** [reply]
I explained it above:

_To make sure you can run tests concurrently and to avoid port usage issues when tests start and stop._
_You can turn off port randomization as well, but you probably also want to turn off test concurrency._

**Reply by ejsmith** [reply]
Hmm... I see that part, but don't see how to turn off the random port assignment.

**Reply by davidfowl** [reply]
[It's not very publicly exposed.](https://github.com/dotnet/aspire/blob/c3d1e743d3da450d6fdaf0ff8ad78cb423e7d9cf/src/Aspire.Hosting.Testing/DistributedApplicationFactory.cs#L161)

@ReubenBond is it possible to set this from tests.

**Reply by ejsmith** [reply]
Alright, I tried doing `appHost.Configuration["DcpPublisher:RandomizePorts"] = "false";` and that worked.

**Reply by davidfowl** [reply]
@ReubenBond did we make this first class?

**Comment by jhancock-taxa** [comment]
Is there a way to just resolve these directly? We're using OpenIddict and you have to set the issuer and it must be the exactly thing that the issuer responds with, which is the actual URL that it is on with port, not the virtual name: https://accounts. I need https://localhost:3424 (or whatever port was randomly assigned) and I need to keep random assignment.

**Reply by robbash** [reply]
@jhancock-taxa, I use this to specify a port:

1. Disable random ports:

```
var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.MyProj>(
  ["DcpPublisher:RandomizePorts=false"]
);
```

2. Create a launch profile on you `MyProj`. Copy the `http` profile for example and update as per your needs. This will also allow you to pass in different environment variables.

3. Set the launch profile env var (as used below) somewhere before your `DistributedApplicationTestingBuilder.CreateAsync<>()` call.

4. You need to choose the launch profile in your `AppHost`'s `Program.cs`:

```
// Note: it's not a standard MS variable, you can name it however. I just stuck to the prefix...
var lp = Environment.GetEnvironmentVariable("ASPNETCORE_LAUNCH_PROFILE");

builder
    .AddProject<Projects.MyProj>("idm", launchProfileName: lp)
    .WithExternalHttpEndpoints();
```
